### PR TITLE
Setter 2.0 commands

### PR DESCRIPTION
**Pull request by pwittrock**
- [x] make the field name optional
- [x] support field path
- [x] have description per setter
- [x] de-dup setter metadata
- [x] replace partial setters with substitutions that aren't lossy

### CODE REVIEW

**Comment by pwittrock** [comment]
@phanimarupaka

File: cmd/config/internal/commands/cmdset.go [hunk 1]
```diff
@@ -32,18 +34,48 @@ func NewSetRunner(parent string) *SetRunner {
...
+		"use this version of the setter format")
+	c.Flags().MarkHidden("version")
 
 	return r
 }
 
+var setterVersion string
+
+var GetOpenAPIFile = func(args []string) (string, error) {
+	return filepath.Join(args[0], "kustomization"), nil
```
Comment (phanimarupaka) [inline cmd/config/internal/commands/cmdset.go]: Is this kustomization or Kptfile?

File: kyaml/openapi/openapi.go [hunk 2]
```diff
@@ -44,6 +45,49 @@ func SchemaForResourceType(t yaml.TypeMeta) *ResourceSchema {
...
+	y, err := yaml.Parse(string(b))
+	if err != nil {
+		return err
+	}
+	// get the field containing the openAPI
+	f := y.Field("openAPI")
+
+	defs, err := f.Value.String()
+	if err != nil {
+		return err
```
Comment (phanimarupaka) [inline kyaml/openapi/openapi.go]: Should we return error here? What if Kptfile is empty and there are no setters yet. It is erroring out.

**Comment by phanimarupaka** [comment]
@pwittrock I am not able to test create-subst e2e. Looks like it's not tied up completely. Other than that added couple of comments.

File: cmd/config/internal/commands/cmdset.go (hunk 1 above)
Comment (pwittrock) [reply cmd/config/internal/commands/cmdset.go]: kustomization

**Comment by pwittrock** [comment]
> @pwittrock I am not able to test create-subst e2e. Looks like it's not tied up completely. Other than that added couple of comments.

What do you mean by this?

```sh
$ cd kustomize/cmd/config/internal/commands
$ go test -v -run TestCreateSubstitutionCommand
=== RUN   TestCreateSubstitutionCommand
=== RUN   TestCreateSubstitutionCommand/substitution_replicas
--- PASS: TestCreateSubstitutionCommand (0.01s)
    --- PASS: TestCreateSubstitutionCommand/substitution_replicas (0.01s)
PASS
ok      sigs.k8s.io/kustomize/cmd/config/internal/commands      0.213s
```

File: kyaml/openapi/openapi.go (hunk 2 above)
Comment (pwittrock) [reply kyaml/openapi/openapi.go]: good catch, `f` would be nil if there were no OpenAPI definitions.  fixed

**Comment by phanimarupaka** [comment]
> > @pwittrock I am not able to test create-subst e2e. Looks like it's not tied up completely. Other than that added couple of comments.
> 
> What do you mean by this?
> 
> ```shell
> $ cd kustomize/cmd/config/internal/commands
> $ go test -v -run TestCreateSubstitutionCommand
> === RUN   TestCreateSubstitutionCommand
> === RUN   TestCreateSubstitutionCommand/substitution_replicas
> --- PASS: TestCreateSubstitutionCommand (0.01s)
>     --- PASS: TestCreateSubstitutionCommand/substitution_replicas (0.01s)
> PASS
> ok      sigs.k8s.io/kustomize/cmd/config/internal/commands      0.213s
> ```

Not the unit tests. I tried to create a kpt binary and run it on a kpt repo package and the command is not discoverable. Can you please try once.

File: cmd/config/internal/commands/cmdset.go (hunk 1 above)
Comment (phanimarupaka) [reply cmd/config/internal/commands/cmdset.go]: But the kpt binary is looking for file named kustomization and failing. We have decided to put it on the Kptfile correct? Are we now storing it in the file name kustomization ?

**Comment by phanimarupaka** [comment]
The create-setter command is creating setter for setter in the kustomization.yaml file. Create-setter should ignore parsing kustomization.yaml file for creating setter.

File: cmd/config/internal/commands/cmdset.go (hunk 1 above)
Comment (phanimarupaka) [inline cmd/config/internal/commands/cmdset.go]: Should this be kustomization.yaml.

Comment (pwittrock) [reply cmd/config/internal/commands/cmdset.go]: no, the .yaml suffix creates issues.  should follow up on this.

**Comment by pwittrock** [comment]
Manually merging since the prow automation does not appear to be configured correctly

**Pull request merged**